### PR TITLE
Core.py upgraded for multiple servo/esc's

### PR DESCRIPTION
**Pull request by worlds6440**
I've upgraded the core to cater for multiple servos and escs. Its fairly self explanitory but basically we now have a dictionary of servo's which we access via a servo enum. This has had the benefit of the set_neutral() method being upgraded to send the adjustable neutral value rather than the hard coded default set in constructor.

I've run tests on the code directly on my pi and screen (and debugged obvious initial problems). It seems to run stable but obviously not tested on prototype.

### CODE REVIEW

File: Calibration.py [hunk 1]
```diff
@@ -1,18 +1,22 @@
...
+    from ordereddict import OrderedDict
 
-
-class CalibrationMode(Enum):
-    CM_NONE = 0
-    CM_LEFT = 1
-    CM_RIGHT = 2
-    CM_LEFT_AUX_1 = 3
-    CM_RIGHT_AUX_1 = 4
+# class CalibrationMode(Enum):
```
Comment (robberwick) [inline Calibration.py]: Is this no longer needed? If not we should remove it. If it is, then it's commented out and won't work :-)

Comment (worlds6440) [reply Calibration.py]: Should be removed. Missed that one.

There were several areas removed due to improved handling of servo's in dictionary meant lots of stuff was no longer required to test which motor/servo was being dealt with.

File: Calibration.py [hunk 2]
```diff
@@ -26,44 +30,52 @@ def __init__(self, core_module, wm, oled):
...
+        # Default current mode to NONE
+        self.mode = ServoEnum.LEFT_MOTOR_ESC
+
+    def get_next_mode(self, mode):
+        """ Find the previous menu item """
+        mode_index = self.menu_list.keys().index(mode)
+        next_index = mode_index + 1
+        if next_index >= len(self.menu_list):
+            next_index = 0  # Wrapped round to end
+        return self.menu_list.keys()[next_index]
```
Comment (robberwick) [inline Calibration.py:48]: I think you can ditch all this manual index checking and just let python raise an exception and catch it:
```
try:
    next_mode = self.menu_list.keys()[mode_index + 1]
except IndexError:
    next_mode = self.menu_list.keys()[0]
return next_mode
```
It's better to ask for forgiveness than permission, ;-)

File: Calibration.py [hunk 3]
```diff
@@ -26,44 +30,52 @@ def __init__(self, core_module, wm, oled):
...
+
+    def get_previous_mode(self, mode):
+        """ Find the previous menu item """
+        mode_index = self.menu_list.keys().index(mode)
+        previous_index = mode_index - 1
+        if previous_index < 0:
+            previous_index = len(self.menu_list) - 1  # Wrapped round to end
```
Comment (robberwick) [inline Calibration.py:55]: I don't think you need to do the manual checking here either. if you index is 0 and you subtract 1, then `my_list[-1]` will return you the last item in the list (negative indexing FTW)
```
In [1]: l = ['one', 'two', 'three']

In [2]: l[0]
Out[2]: 'one'

In [3]: l[-1]
Out[3]: 'three'
```

so you can just get away with returning `self.menu_list.keys()[previous_index]` immediately

File: Calibration.py [hunk 4]
```diff
@@ -26,44 +30,52 @@ def __init__(self, core_module, wm, oled):
...
-                title = "Right Aux 1:"
-                message = str(self.core.right_aux_1_servo.servo_min) + '/'\
-                    + str(self.core.right_aux_1_servo.servo_mid) + '/'\
-                    + str(self.core.right_aux_1_servo.servo_max)
-
-            self.oled.cls()  # Clear Screen
+            servo = self.core.servos[servo_id]
+            title = "{}:".format(servo[2])
+            message = "{} / {} / {}".format(
+                str(servo[0].servo_min),
```
Comment (robberwick) [inline Calibration.py:64]: I think the `str()` calls are redundant here, as you are already string formatting.

File: Calibration.py [hunk 5]
```diff
@@ -153,35 +142,22 @@ def read_config(self):
...
-                self.core.right_aux_1_servo.servo_max = \
-                    int(config.get('motors', 'RIGHT_AUX_1_MAX'))
+                # Read in min/mid/max of menu items that we can change.
+                for item in self.menu_list.items():
+                    item_enum_str = str(item[0]).split('.')
+                    item_min_str = "{}_MIN".format(item_enum_str[0])
+                    item_mid_str = "{}_MID".format(item_enum_str[0])
+                    item_max_str = "{}_MAX".format(item_enum_str[0])
+                    try:
+                        self.core.servos[item[0]][0].servo_min = \
```
Comment (robberwick) [inline Calibration.py:144]: Lots of accessing of multi-dimensional arrays in this way makes me very uneasy. It is an approach that makes it *very* easy to ntroduce bugs, particularly when the code is being picked up by a future dev. I'm not sure what the answer is here, but generally, replacing an approach which relies on accessing a list with an index with one which uses a dict and named keys, or even a plain object with properties would be preferable.

In short, it's probably fine for now, but it's certainly technical debt.

File: Calibration.py [hunk 6]
```diff
@@ -200,41 +176,24 @@ def write_config(self):
...
+                item_min_str,
+                str(self.core.servos[item[0]][0].servo_min))
+            config.set(
+                'motors',
+                item_mid_str,
+                str(self.core.servos[item[0]][0].servo_mid))
+            config.set(
+                'motors',
+                item_max_str,
+                str(self.core.servos[item[0]][0].servo_max))
```
Comment (robberwick) [inline Calibration.py:188]: I think you can ditch some of this boilerplate with judicious use of `zip`, and a bit of `getattr`:
```
for item in self.menu_list.items():
    item_enum_str = str(item[0]).split('.')[0]
    for (suffix, prop) in zip(('MIN', 'MID', 'MAX'), ('servo_min', 'servo_mid', 'servo_max')):
        config_set(
            'motors',
            '{0}_{1}'.format(item_enum_str, suffix),
            str(getattr(self.core.servos[item[0]][0], prop))
        )
```

File: core.py [hunk 7]
```diff
@@ -24,49 +50,52 @@ class Core():
...
-
-        # Proximity sensor: roughly cm from closest measurable point
-        # self.tof_left = sensor.Sensor(0, 450, 20, 0)
-        # self.PWMservo = PWM.Servo(pulse_incr_us=1)
+        # Create a list of servo's
+        self.servos = dict()
+        # Add Motor Servo's. NOTE: Left motor esc is reversed.
+        self.servos[ServoEnum.LEFT_MOTOR_ESC] = [
+            servo_control.Servo_Controller(
+                800, 1300, 1800, True  # Min, Mid, Max, Reversed
```
Comment (robberwick) [inline core.py]: suggest using name paramsfor these rather than positional arguments. They're easier to parse and less likely to suffer from rot that accompanying comments.

File: core.py [hunk 8]
```diff
@@ -98,46 +129,70 @@ def enable_motors(self, enable):
...
-        left_micros = self.left_servo.micros(left_speed)
-        right_micros = self.right_servo.micros(right_speed)
+        left_micros = 0
+        right_micros = 0
+        try:
+            if left_servo != ServoEnum.SERVO_NONE:
+                left_micros = self.servos[left_servo][0].micros(left_speed)
+            if right_servo != ServoEnum.SERVO_NONE:
+                right_micros = self.servos[right_servo][0].micros(right_speed)
+        except:
```
Comment (robberwick) [inline core.py:148]: Is a wide open `except` here what you want? They're usually a warning flag that you'll end up catching exceptions that perhaps you may actually want to raise higher in the stack. What failure mode are you expecting?

File: core.py [hunk 9]
```diff
@@ -98,46 +129,70 @@ def enable_motors(self, enable):
...
+                try:
+                    if left_servo != ServoEnum.SERVO_NONE:
+                        self.PWMservo.set_servo(
+                            self.servos[left_servo][1],
+                            left_micros)
+                    if right_servo != ServoEnum.SERVO_NONE:
+                        self.PWMservo.set_servo(
+                            self.servos[right_servo][1],
+                            right_micros)
+                except:
```
Comment (robberwick) [inline core.py:166]: Again, is broad `except` here what you want?

File: core.py [hunk 10]
```diff
@@ -98,46 +129,70 @@ def enable_motors(self, enable):
...
+    def set_neutral(self,
+                    left_servo=ServoEnum.LEFT_MOTOR_ESC,
+                    right_servo=ServoEnum.RIGHT_MOTOR_ESC):
+        """ Send neutral to the motors IMEDIATELY. """
+        try:
+            if left_servo != ServoEnum.SERVO_NONE:
+                left_mid = self.servos[left_servo][0].servo_mid
+            if right_servo != ServoEnum.SERVO_NONE:
+                right_mid = self.servos[right_servo][0].servo_mid
+        except:
```
Comment (robberwick) [inline core.py:178]: Same question re: broad `except`

File: core.py [hunk 11]
```diff
@@ -98,46 +129,70 @@ def enable_motors(self, enable):
...
+            try:
+                if left_servo != ServoEnum.SERVO_NONE:
+                    self.PWMservo.set_servo(
+                        self.servos[left_servo][1],
+                        left_mid)
+                if right_servo != ServoEnum.SERVO_NONE:
+                    self.PWMservo.set_servo(
+                        self.servos[right_servo][1],
+                        right_mid)
+            except:
```
Comment (robberwick) [inline core.py:194]: and another.

File: core.py [hunk 12]
```diff
@@ -149,10 +204,6 @@ def read_sensor(self, pin):
...
     def stop(self):
-        if self.arduino:
-            self.arduino.direct_micros(self.LEFT_MID, self.RIGHT_MID)
-        else:
-            self.PWMservo.set_servo(LEFT_SERVO_PIN, self.LEFT_MID)
-            self.PWMservo.set_servo(RIGHT_SERVO_PIN, self.RIGHT_MID)
-            for pin in range(0, 3):
-                i2c_lidar.turnoff(LIDAR_PINS[pin])
+        self.set_neutral()
+        for pin in range(0, 3):
```
Comment (robberwick) [inline core.py:208]: is `LIDAR_PINS` the list of pins? Then you can:
```
for pin in LIDAR_PINS:
    i2c_lidar.turnoff(LIDAR_PINS[pin])
```
or, for bonus single line points:
`[i2c_lidar.turnoff(LIDAR_PINS[pin]) for pin in LIDAR_PINS]`

File: servo_control.py [hunk 13]
```diff
@@ -4,6 +4,7 @@
 class Servo_Controller():
 
     def __init__(self, min, mid, max, bReverse):
```
Comment (robberwick) [inline servo_control.py]: Use named params here and define default values to use as backup in case they're not supplied?

**Review by robberwick — COMMENTED** [review]
A few minor comments, a couple of questions, and soem suggestions for brevity through increased pythonification.

File: core.py (hunk 8 above)
Comment (worlds6440) [reply core.py:148]: Wasn't sure exactly what exception to catch. How would I find what it throws?

File: Calibration.py (hunk 5 above)
Comment (worlds6440) [reply Calibration.py:144]: That was my thought at midnight. Didn't like it, but its tricky to see clearly how the dict inside dict would be accessed.

File: core.py (hunk 7 above)
Comment (worlds6440) [reply core.py]: ok, good call

File: servo_control.py (hunk 13 above)
Comment (worlds6440) [reply servo_control.py]: again, good call.

File: core.py (hunk 8 above)
Comment (robberwick) [reply core.py:148]: I guess that depends on what failure mode you're expecting. `self.servos[left_servo]` would fail with an `IndexError`, or if you *do* happen to get something back, but it doesn't have a `micros` property, then that would fail with an `AttributeError`.

What were you attempting to trap out here? Because at the moment, your calling code doesn't know that there was a failure, because all you're doing is `print()`ing and then continuing (i.e. returning `None`). Another thing to consider is either not handling the error here, and catching it in the calling code, or handling it, and raising a custom Exception which might be more meaningful in the calling code?

File: Calibration.py (hunk 3 above)
Comment (worlds6440) [reply Calibration.py:55]: Fair do, just feels wrong to me :)

File: core.py [hunk 14]
```diff
@@ -80,16 +106,18 @@ def __init__(self, tof_lib):
...
-            for pin in range(0, 3):
-                i2c_lidar.xshut([LIDAR_PINS[pin]])
-                self.lidars.append(
-                    i2c_lidar.create(LIDAR_PINS[pin], tof_lib, 0x2a + pin)
-                )
+
+        # I2C lidar sensors should be enabled
+        # in both Arduino mode or direct pi mode.
+        for pin in range(0, 3):
+            i2c_lidar.xshut([LIDAR_PINS[pin]])
```
Comment (PKM-UK) [inline core.py:116]: This was the cause of the sensor issues last week; this should call i2c_lidar.xshut(LIDAR_PINS) before iterating over the pins turning them on and assigning addresses.

File: Calibration.py (hunk 3 above)
Comment (robberwick) [reply Calibration.py:55]: you're still thinking in C++, negative indexing is totally legit in python. ;-)

Just wait until you bump in to slicing notation. The optional step parameter will blow your mind:

```
l = [1, 2, 3, 4, 5]
l[1] => 2
l[1:] => [2, 3, 4, 5]
l[:1] => [1]
l[::2] => [1, 3, 5]
l[::-2] => [5, 3, 1]
```
TL;DR: Python is *awesome* with lists ;-)

File: core.py (hunk 14 above)
Comment (robberwick) [reply core.py:116]: also as per [above comment ](https://github.com/hackhitchin/piwars3/pull/17#discussion_r107375706),there's no need for the additional `for..in range` construct. Just jump straight in with a `[i2c_lidar.xshut(LIDAR_PINS[pin]) for pin in LIDAR_PINS]` to iterate the list assigning pins.

Comment (PKM-UK) [reply core.py:116]: You don't even need that "for pin in LIDAR_PINS": the xshut function takes a list and iterates internally so you can just call it with LIDAR_PINS as the argument, no iteration needed here at all.

These functions have changed in meaning since they were first written and we ditched the XSHUT pin so could do with being renamed/refactored at some point, but as long as the sensors work I can't get worked up about it.

Comment (robberwick) [reply core.py:116]: well, that's even better! Also, if the function is now misnamed, then *definitely* change it to something more meaningful or future travellers will be all kinds of confused. It'll only take a minte and future us will be grateful.
